### PR TITLE
unique listener metrics

### LINKED ISSUE #415: Implement Unique Listener Metrics
There are some concerns with the data as represented in the design file, but I think we can build this using existing elements within Metrics.

*Note the sidenav menu doesn't need to be implemented with this design. But we do need to include a link to the Unique Listener Metric dashboard.

Sketch file: https://sketch.cloud/s/y03oq

Original design ticket:
https://github.com/PRX/metrics.prx.org/issues/413#issuecomment-557206981

### LINKED ISSUE #419: google sheets exports data as strings
From Ma'ayan:

> I did a Sheets export from Metrics and was trying to figure out why basic things like `=average(cell:cell)` formulas were returning errors and I found that each cell had an invisible ’ (that is, until you clicked on the cell) at the start of each number.

### PR DESCRIPTION
**Pull request by sandikbarr**
closes #415 
Adds Unique Listeners feature with line chart and data export.

closes #419 
Fixes google sheets export for numeric data (was exporting all data as strings 😬  because I had just reused the CSV export code.)

Prettier really got a hold of some of these files. 😐

### CODE REVIEW

File: src/app/ngrx/reducers/selectors/export.selectors.ts [hunk 1]
```diff
@@ -22,85 +46,110 @@ import { selectRoutedGroupCharted } from './group-charted.selectors';
...
+    return [
+      {
+        label: 'Unique Listeners',
+        data: routedPodcastData.listeners
+      }
+    ];
+  }
+});
+
+export const toCsvArray = (downloads: ExportData[], dateFormat?: Function): (string | number)[][] => {
```
Comment (sandikbarr) [inline src/app/ngrx/reducers/selectors/export.selectors.ts:152]: fixes #419 by removing toString() from numeric data here and using `(string | number)[]`

File: src/app/shared/menu/menu-bar.component.ts [hunk 2]
```diff
@@ -1,58 +1,57 @@
...
-        <metrics-interval-dropdown [routerParams]="routerParams$ | async"></metrics-interval-dropdown>
-        <div class="separator"></div>
-        <ng-container *ngIf="(metricsType$ | async) !== dropday; else dropdayMenu">
-          <metrics-standard-date-range-dropdown [interval]="interval$ | async" [standardRange]="standardRange$ | async">
-          </metrics-standard-date-range-dropdown>
-          <metrics-custom-date-range-dropdown [routerParams]="routerParams$ | async"></metrics-custom-date-range-dropdown>
-        </ng-container>
-        <ng-template #dropdayMenu>
-          <metrics-days-dropdown [routerParams]="routerParams$ | async"></metrics-days-dropdown>
-        </ng-template>
```
Comment (sandikbarr) [inline src/app/shared/menu/menu-bar.component.ts:32]: cleaned this up a little with less `| async` subscriptions

File: src/app/ngrx/reducers/selectors/export.selectors.ts [hunk 3]
```diff
@@ -22,85 +46,110 @@ import { selectRoutedGroupCharted } from './group-charted.selectors';
...
   }
 );
 
-export const toCsvArray = (downloads: ExportData[], dateFormat?: Function): string[][] => {
+export const selectExportListeners = createSelector(selectRoutedPodcastListeners, (routedPodcastData: PodcastListeners) => {
+  if (routedPodcastData && routedPodcastData.listeners) {
+    return [
+      {
+        label: 'Unique Listeners',
+        data: routedPodcastData.listeners
```
Comment (cavis) [inline src/app/ngrx/reducers/selectors/export.selectors.ts:146]: I find it a bit weird that the dates are the x-axis, and all the data ends up on row 2.  But I guess that's the way our other chart exports work (where there are multiple lines).

File: src/app/ngrx/reducers/selectors/export.selectors.ts (hunk 1 above)
Comment (cavis) [reply src/app/ngrx/reducers/selectors/export.selectors.ts:152]: Nice!  Free bugfix!

File: src/app/ngrx/reducers/selectors/export.selectors.ts [hunk 4]
```diff
@@ -112,53 +161,70 @@ export const toCsvArray = (downloads: ExportData[], dateFormat?: Function): stri
...
-    }).join(',')).join('\r\n');
+      data
+        .map(
+          row =>
+            row &&
+            row
+              .map(v => {
+                return v && typeof v === 'string' && v.indexOf(',') > -1 ? `"${v}"` : v;
+              })
+              .join(',')
```
Comment (cavis) [inline src/app/ngrx/reducers/selectors/export.selectors.ts:190]: Some ... _interesting_ indentation choices by Prettier here.

File: src/app/downloads/listeners.component.spec.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+      });
+  }));
+
+  it('should show loading spinner when loading', () => {
+    dispatchHelper.dispatchLoadPodcastListeners(store);
+    fix.detectChanges();
+    expect(de.query(By.css('prx-spinner'))).not.toBeNull();
+  });
+
+  it('should show a listeners chart', () => {
```
[the remark this answers is not in the document]
Comment (sandikbarr) [reply src/app/downloads/listeners.component.spec.ts:55]: yes, that. There isn't much to test in the component yet, but always good to start with a basic spec setup before it gets complicated.

File: src/app/ngrx/reducers/models/interval.model.ts [hunk 6]
```diff
@@ -4,8 +4,10 @@ export interface IntervalModel {
   key: string;
 }
 
+export const INTERVAL_LASTWEEK: IntervalModel = { value: 'LAST_WEEK', name: 'Last 7 Days', key: 'last_week' };
+export const INTERVAL_LAST28DAYS: IntervalModel = { value: 'LAST_28', name: 'Last 28 Days', key: 'last_28' };
```
[the remark this answers is not in the document]
Comment (sandikbarr) [reply src/app/ngrx/reducers/models/interval.model.ts:8]: Yeah, I like these labels. That helps. I'm going to update this by making calendar week and calendar month separate intervals and stop trying to re-use the existing intervals.

**Pull request merged**